rpmdep.1
.\" Automatically generated by Pod::Man 2.25 (Pod::Simple 3.16)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "RPMDEP 1"
.TH RPMDEP 1 "2012-08-01" "perl v5.14.2" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
rpmdep \- display the full dependency of an installed rpm package
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
rpmdep search recursively for package dependencies. 
It resolvs all dependencies to package names.
It can also prepare a file to build a graph of dependencies, with graphviz.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
rpmdep.pl  [options] package
.PP
options:
.PP
.Vb 5
\&   \-help                brief help message
\&   \-man                 full documentation
\&   \-V, \-\-version        print version
\&   \-use\-cache           use cache file instead rpm query
\&   \-clear\-cache         clear cache file
\&
\&   \-verbose             verbose
\&   \-dot dotfile         build a dot file for graphviz
.Ve
.SH "REQUIRED ARGUMENTS"
.IX Header "REQUIRED ARGUMENTS"
the package to resolv
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-help\fR" 8
.IX Item "-help"
Print a brief help message and exits.
.IP "\fB\-man\fR" 8
.IX Item "-man"
Print the manual page and exits.
.IP "\fB\-version\fR" 8
.IX Item "-version"
Print the program release and exit.
.IP "\fB\-verbose\fR" 8
.IX Item "-verbose"
The program works and print debugging messages.
.IP "\fB\-dot\fR" 8
.IX Item "-dot"
create a dot file to be used by graphviz
.IP "\fB\-use\-cache\fR" 8
.IX Item "-use-cache"
the rpm query may be long (10 to 30 s). If you will run an rpmorphan tool
several time, this option will allow to gain a lot of time :
it save the rpm query on a file cache (first call), then
use this cache instead quering rpm (others calls).
.IP "\fB\-clear\-cache\fR" 8
.IX Item "-clear-cache"
to remove cache file. Can be used with \-use\-cache to write
a new cache.
.SH "USAGE"
.IX Header "USAGE"
rpmdep.pl \-\-use\-cache \-dot bash.dot bash
.PP
dot \-Tps bash.dot \-o bash.ps
.SH "FILES"
.IX Header "FILES"
the program can use the /tmp/rpmorphan.cache file
.SH "DIAGNOSTICS"
.IX Header "DIAGNOSTICS"
the verbose mode allow to see all the recursive work
.PP
others messages are
.IP "\fBpackage ... is not installed\fR" 8
.IX Item "package ... is not installed"
this is not a name of an installed package
.IP "\fB... has no dependencies\fR" 8
.IX Item "... has no dependencies"
this can comes for some admin tools (ash for example)
.IP "\fB... depends upon (list)\fR" 8
.IX Item "... depends upon (list)"
return a list of sorted and comma separated packages
.SH "EXIT STATUS"
.IX Header "EXIT STATUS"
O if all is ok
.PP
>=1 if a problem
.SH "CONFIGURATION"
.IX Header "CONFIGURATION"
nothing
.SH "DEPENDENCIES"
.IX Header "DEPENDENCIES"
you should use graphviz to build graph from
.SH "INCOMPATIBILITIES"
.IX Header "INCOMPATIBILITIES"
not known
.SH "BUGS AND LIMITATIONS"
.IX Header "BUGS AND LIMITATIONS"
the program does not work well on program installed
with several versions
.SH "NOTES"
.IX Header "NOTES"
this program can be used as \*(L"normal\*(R" user
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\fIrpm\fR\|(1) for rpm call
.PP
\fIrpmorphan\fR\|(1)
.PP
\fIrpmusage\fR\|(1)
.PP
\fIrpmduplicates\fR\|(1)
.PP
\fIrpmextra\fR\|(1)
.SH "LICENSE AND COPYRIGHT"
.IX Header "LICENSE AND COPYRIGHT"
Copyright (C) 2008 by Eric Gerbier
This program is free software; you can redistribute it and/or modify
it under the terms of the \s-1GNU\s0 General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.SH "AUTHOR"
.IX Header "AUTHOR"
Eric Gerbier
.PP
you can report any bug or suggest to gerbier@users.sourceforge.net


NEWS
			1.11 version
			-------------
new features :
* add suggests option
* add /etc/rpmorphanrc
* add env RPMORPHAN_METHOD for debugging

bugfix :
* fix bug with space in filenames
-------------------------------------------------------------------------------
			1.10 version
			-------------
new features :
* display package summary in a popup on S key

bugfix :
* bugfix eval for Curses (mandriva)
* fix bad version display

improvment :
* reduce size/duplicated code in rpmorphan/rpmusage
* reduce menu length in curses gui (< 72)

others :
* clean all scripts
-------------------------------------------------------------------------------
			1.9 version
			-------------
new features :
- add rpmextra tool
- add grpmorphan as link to rpmorphan -gui
others :
- fix perlcritic warnings
-------------------------------------------------------------------------------
			1.8 version
			-------------
bugfix : 
- fix bug on rpmdep ( missing installtime in urpm mode)

new features :
- add help translation (gui)
-------------------------------------------------------------------------------
			1.7 version
			-------------
bugfix : 
- Curses is no more an rpm dependency

new features : 
- add reload button
-------------------------------------------------------------------------------
			1.6 version
			-------------
other :
- add a confirm dialog box on remove action
- curses window size is now variable to use full available size
- info window is now scrollable (curses gui)
-------------------------------------------------------------------------------
			1.5 version
			-------------
new features : 
- use URPM perl package for faster database access on mandriva
- can use RPM2 perl package for faster database access (fedora)
bugfix : 
- fix a bug on curses interface
- remove all matching packages
others :
- merge spec file with fedora one
- add info for fedora/mandriva users
-------------------------------------------------------------------------------
			1.4 version
			-------------
other :
when removing a package, anf if rpm send some output, the code now check the rpm 
database to know if the remove works or not
-------------------------------------------------------------------------------
			1.3 version
			-------------
incompatibility : none
bugfix : 
- fix a bug if exclude are set in config file (thanks Szymon Siwek)
new features : 
- display number of deleted file
other :
-------------------------------------------------------------------------------
			1.2 version
			-------------
incompatibility : none
bugfix : none
new features : 
- write log to /var/log/rpmorphan.log
- add rpmdep.pl tool
- add rpmduplicate.pl tool
other :
-------------------------------------------------------------------------------
			1.1 version
			-------------
incompatibility : none
bugfix : none
new features : 
- (rpmorphan) add a default target : guess-lib
- (rpmorphan) add keep-file option, which allow to cutomize the keep file
- (rpmusage) add a default target : all
other :
- the different tools now share a common library
- rpmlint was used to improve the rpm package

-------------------------------------------------------------------------------
			1.0 version
			-------------
add dry-run (simulation) mode
(gui) add 3 counters : packages, orphans, selected
(gui) recursive analysis
improve virtuals : now check if how many package provide this one
-------------------------------------------------------------------------------
			0.9 version
			-------------
add another graphical interface (curses), available by the --curses command line
argument
add configuration file
-------------------------------------------------------------------------------
			0.8 version
			-------------
add simple graphical user interface (optionnal), which require
the perl-tk package :use --gui command line argument

the code has been split in smaller subroutines.
-------------------------------------------------------------------------------
			0.4 version
			-------------
add optionnal cache, which can be shared by all rpmorphan tools
to spare rpm query time :
use --use-cache and --clear-cache command line arguments
-------------------------------------------------------------------------------
			0.3 version
			-------------
new feature : add rpmusage tool
-------------------------------------------------------------------------------

			0.2 version
			-------------
new feature : add a permanent list of exclude file (default is empty) and
4 new options for rpmorphan : --zero-keep, --list-keep, --add-keep, --del-keep

-------------------------------------------------------------------------------
			0.1 version
			-------------

first public release


rpmdep.pl
#!/usr/bin/perl 
###############################################################################
#   rpmdep.pl
#
#    Copyright (C) 2006 by Eric Gerbier
#    Bug reports to: gerbier@users.sourceforge.net
#    $Id: rpmorphan-1.11 | rpmdep.pl | Mon Jul 4 12:46:50 2011 +0000 | gerbier  $
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
###############################################################################
# this program search the full dependency of an installed rpm package
###############################################################################
use strict;
use warnings;

use English '-no_match_vars';

use Getopt::Long;    # arg analysis
use Pod::Usage;      # man page
use Data::Dumper;    # debug

# library
use File::Basename;
my $dirname = dirname($PROGRAM_NAME);
require $dirname . '/rpmorphan-lib.pl'
  or die "problem to load rpmorphan-lib: $ERRNO\n";

# the code should use no "print" calls
# but instead debug, warning, info calls
#########################################################
# is to be defined because rpmorphan-lib need it
sub display_status($) {
	debug( shift @_ );
	return;
}
##########################################################
# resolv problems with characters in names for graphwiz
sub norm($) {
	my $name = shift @_;

	my $norm_name = $name;
	$norm_name =~ s/-/_/g;
	$norm_name =~ s/\./_/g;
	$norm_name =~ s/\+//g;

	return $norm_name;
}
#########################################################
# search the dependency of $name
# no prototype because recursive
sub solve {
	my $name        = shift @_;    # package name
	my $rh_provides = shift @_;    # general provide hash
	my $rh_depends  = shift @_;    # general dependencies hash
	my $rh_files    = shift @_;    # general dependencies hash
	my $rh_res      = shift @_;    # result cache
	my $fh_dot      = shift @_;    # flag for dot
	my $level       = shift @_;    # recurse level

	$level++;

	my $state = 0;                 # state ( 0 for ok)

	foreach my $dep ( keys %{ $rh_depends->{$name} } ) {
		my $debug = "(solve $level) $name -> $dep ";

		# dep may be a package or a file or a virtual
		my $pac;                   # searched package

		# analyse and resolv dependency
		if ( exists $rh_files->{$dep} ) {
			debug("$dep is a package");
			$pac = $dep;
		}
		elsif ( $dep =~ m/^rpmlib/ ) {

			# strange case, but real, for example :
			# rpmlib(PayloadFilesHavePrefix)
			# rpmlib(CompressedFileNames)
			debug( $debug . ' (skip rpmlib)' );
			next;
		}
		elsif ( exists $rh_provides->{$dep} ) {
			debug("$dep is a file");
			$pac = $rh_provides->{$dep};
			$debug .= " -> $pac ";

			#debug($debug);
		}
		else {
			warning("can not find who provide $dep");
			$state++;
			next;
		}

		# filter dependency
		if ( exists $rh_res->{$pac} ) {

			# already found
			debug( $debug . ' (already found)' );
		}
		elsif ( $pac eq $name ) {

			# no dependency on self
			debug( $debug . ' (not on self)' );
		}
		else {
			if ($fh_dot) {
				print {$fh_dot} norm($name) . ' -> ' . norm($pac) . ";\n";
			}
			$rh_res->{$pac} = 1;
			debug($debug);
			my $sub_state =
			  solve( $pac, $rh_provides, $rh_depends, $rh_files, $rh_res,
				$fh_dot, $level );
			$state += $sub_state;
		}

	}    # foreach
	return $state;
}
#########################################################
#
#	main
#
#########################################################
my $version = '0.5';

my $opt_help;
my $opt_man;
my $opt_version;
my $opt_use_cache;
my $opt_clear_cache;
my $opt_verbose;

my $opt_dot;

my %opt = (
	'help'        => \$opt_help,
	'man'         => \$opt_man,
	'verbose'     => \$opt_verbose,
	'version'     => \$opt_version,
	'dot'         => \$opt_dot,
	'use-cache'   => \$opt_use_cache,
	'clear-cache' => \$opt_clear_cache,
);

Getopt::Long::Configure('no_ignore_case');
GetOptions( \%opt, 'help|?', 'man', 'verbose', 'version|V', 'dot=s',
	'use-cache!', 'clear-cache' )
  or pod2usage(2);

init_debug($opt_verbose);

if ($opt_help) {
	pod2usage(1);
}
elsif ($opt_man) {
	pod2usage( -verbose => 2 );
}
elsif ($opt_version) {
	print_version($version);
	exit;
}

# test if a target is set
if ( $#ARGV != 0 ) {
	pod2usage('need a target : package name');
}
my $package = $ARGV[0];

# hash structures to be filled with rpm query
my %files;
my %install_time;
my %provides;
my %depends;
my %virtual;
my %requires;

read_rpm_data( \%opt, \%provides, \%install_time, \%files, \%depends, \%virtual,
	\%requires );

if ( !exists $install_time{$package} ) {
	warning("package $package is not installed");
	exit 1;
}

# first we have to change order used %depends
my %depends_from;
foreach my $key1 ( keys %depends ) {
	foreach my $key2 ( keys %{ $depends{$key1} } ) {

		# ex key2 = afick-gui key1 = afick
		$depends_from{$key2}{$key1} = 1;
	}
}

# same for provides
my %is_provided_by;
foreach my $key1 ( keys %provides ) {
	foreach my $key2 ( @{ $provides{$key1} } ) {

		# key2 is provided by $key1 package
		$is_provided_by{$key2} = $key1;
	}
}

# dot graph ?
my $fh_dot;
if ($opt_dot) {

	# open output file
	## no critic (RequireBriefOpen)
	if ( open $fh_dot, '>', $opt_dot ) {
		print {$fh_dot} "digraph \"rpmdep\" {\n";
	}
	else {
		warning("can not write to $opt_dot file : $ERRNO\n");
		$opt_dot = 0;
	}
}

debug('2 : solve');
my %res;
my $state =
  solve( $package, \%is_provided_by, \%depends_from, \%files, \%res, $fh_dot,
	0 );

my $res = join q{,}, sort keys %res;
if ($res) {
	info("$package depends upon $res");
}
else {
	info("$package has no dependencies");
}

if ($opt_dot) {
	print {$fh_dot} "}\n";
	close $fh_dot or die "problem to close dot file :$OS_ERROR\n";
}

exit $state;
__END__

=head1 NAME

rpmdep - display the full dependency of an installed rpm package

=head1 DESCRIPTION

rpmdep search recursively for package dependencies. 
It resolvs all dependencies to package names.
It can also prepare a file to build a graph of dependencies, with graphviz.

=head1 SYNOPSIS

rpmdep.pl  [options] package

options:

   -help                brief help message
   -man                 full documentation
   -V, --version        print version
   -use-cache		use cache file instead rpm query
   -clear-cache		clear cache file

   -verbose             verbose
   -dot dotfile		build a dot file for graphviz

=head1 REQUIRED ARGUMENTS

the package to resolv

=head1 OPTIONS

=over 8

=item B<-help>

Print a brief help message and exits.

=item B<-man>

Print the manual page and exits.

=item B<-version>

Print the program release and exit.

=item B<-verbose>

The program works and print debugging messages.

=item B<-dot>

create a dot file to be used by graphviz

=item B<-use-cache>

the rpm query may be long (10 to 30 s). If you will run an rpmorphan tool
several time, this option will allow to gain a lot of time :
it save the rpm query on a file cache (first call), then
use this cache instead quering rpm (others calls).

=item B<-clear-cache>

to remove cache file. Can be used with -use-cache to write
a new cache.

=back

=head1 USAGE

rpmdep.pl --use-cache -dot bash.dot bash

dot -Tps bash.dot -o bash.ps

=head1 FILES

the program can use the /tmp/rpmorphan.cache file

=head1 DIAGNOSTICS

the verbose mode allow to see all the recursive work

others messages are

=over 8

=item B<package ... is not installed>

this is not a name of an installed package

=item B<... has no dependencies>

this can comes for some admin tools (ash for example)

=item B<... depends upon (list)>

return a list of sorted and comma separated packages

=back

=head1 EXIT STATUS

O if all is ok

>=1 if a problem

=head1 CONFIGURATION

nothing

=head1 DEPENDENCIES

you should use graphviz to build graph from 

=head1 INCOMPATIBILITIES

not known

=head1 BUGS AND LIMITATIONS

the program does not work well on program installed
with several versions

=head1 NOTES

this program can be used as "normal" user

=head1 SEE ALSO

=for man
\fIrpm\fR\|(1) for rpm call
.PP
\fIrpmorphan\fR\|(1)
.PP
\fIrpmusage\fR\|(1)
.PP
\fIrpmduplicates\fR\|(1)
.PP
\fIrpmextra\fR\|(1)

=for html
<a href="rpmorphan.1.html">rpmorphan(1)</a><br />
<a href="rpmusage.1.html">rpmusage(1)</a><br />
<a href="rpmduplicates.1.html">rpmduplicates(1)</a><br />
<a href="rpmextra.1.html">rpmextra(1)</a><br />

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2008 by Eric Gerbier
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

=head1 AUTHOR

Eric Gerbier

you can report any bug or suggest to gerbier@users.sourceforge.net

=cut

